### PR TITLE
Implement fcvt_to_uint_sat (f32x4 -> i32x4) for x86

### PR DESCRIPTION
**Pull request by abrown**
~~This follows from #1821 and should be merged after that PR.~~ [ready to review]

Along the lines of #1820, this adds the necessary legalizations and instructions for enabling the Wasm SIMD spec tests with truncation instructions.

### CODE REVIEW

File: cranelift/codegen/src/isa/x86/enc_tables.rs [hunk 1]
```diff
@@ -1219,6 +1219,92 @@ fn expand_fcvt_to_uint_sat(
     cfg.recompute_block(pos.func, done);
 }
 
+// Lanes of an I32x4 filled with the max signed integer values converted to an F32x4. TODO flip?
+static MAX_SIGNED_I32X4S_AS_F32X4S: [u8; 16] = [
+    0x4f, 0x00, 0x00, 0x00, 0x4f, 0x00, 0x00, 0x00, 0x4f, 0x00, 0x00, 0x00, 0x4f, 0x00, 0x00, 0x00,
+];
+
+/// This legalization converts a vector of 32-bit floating point lanes to signed integer lanes
+/// using CVTTPS2DQ (see encoding of `x86_cvtt2si`). If user-defined flags `assert_no_nans` and
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x86/enc_tables.rs:1228]: I thought the `assert_no_nans` idea had been abandoned.  And what is `assert_in_bounds` ?

File: cranelift/codegen/src/isa/x86/enc_tables.rs [hunk 2]
```diff
@@ -1219,6 +1219,92 @@ fn expand_fcvt_to_uint_sat(
...
+    {
+        let controlling_type = pos.func.dfg.ctrl_typevar(inst);
+        if controlling_type == I32X4 {
+            debug_assert_eq!(pos.func.dfg.value_type(arg), F32X4);
+            let x86_isa = isa
+                .as_any()
+                .downcast_ref::<isa::x86::Isa>()
+                .expect("the target ISA must be x86 at this point");
+            if x86_isa.isa_flags.use_avx512vl_simd() || x86_isa.isa_flags.use_avx512f_simd() {
+                // When no NaNs are possible and we know the results do not need saturation, we can
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x86/enc_tables.rs:1256]: Is this comment current?  If yes, is it the case that `x86_isa.isa_flags.use_avx512vl_simd() || x86_isa.isa_flags.use_avx512f_simd()` implies "No NaNs are possible and results don't need saturation" ?

File: cranelift/codegen/src/isa/x86/enc_tables.rs (hunk 1 above)
Comment (abrown) [reply cranelift/codegen/src/isa/x86/enc_tables.rs]: Sorry, must have missed removing these comments; they are gone now.

File: cranelift/codegen/src/isa/x86/enc_tables.rs (hunk 2 above)
Comment (abrown) [reply cranelift/codegen/src/isa/x86/enc_tables.rs]: Same as above... fixed.

File: cranelift/codegen/src/isa/x86/enc_tables.rs [hunk 3]
```diff
@@ -1219,6 +1219,90 @@ fn expand_fcvt_to_uint_sat(
...
+    } = pos.func.dfg[inst]
+    {
+        let controlling_type = pos.func.dfg.ctrl_typevar(inst);
+        if controlling_type == I32X4 {
+            debug_assert_eq!(pos.func.dfg.value_type(arg), F32X4);
+            let x86_isa = isa
+                .as_any()
+                .downcast_ref::<isa::x86::Isa>()
+                .expect("the target ISA must be x86 at this point");
+            if x86_isa.isa_flags.use_avx512vl_simd() || x86_isa.isa_flags.use_avx512f_simd() {
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x86/enc_tables.rs:1254]: Are the rounding semantics of the two alternative implementations equivalent?  For the short case, `vcvtps2udq`, the Intel manual says "When a conversion is inexact, the value returned is rounded according to the rounding control bits in the MXCSR register or the embedded rounding control bits."  The slow path appears to use `cvtt2si` (presumably, really, `cvttps2pi`), and all the `tt` variant conversion insns are round-towards-zero: "When a conversion is inexact, a truncated (round toward zero) result is returned."

Comment (abrown) [reply cranelift/codegen/src/isa/x86/enc_tables.rs:1348]: This is a good point: I had hoped MXCSR would have the right rounding mode but it does not. I added a `run` filetest that passes with the long sequence and fails with the AVX512 sequence: it looks like the machines running the CI have AVX512 and therefore fail. I started playing around with a new recipe that embeds the rounding control (e.g. `EvexContext::RoundingRegToRegFP { rc: EvexRoundingControl::RZ }` but this didn't immediately fix the test case (the emitted bytes aren't exactly right) so I will have to play around with this a bit more.

**Comment by abrown** [comment]
Replacing this with #1990--I propose we look at the AVX512 lowering in a separate PR since it may get tricky (and this legalization is already tricky enough).

**Pull request closed without merging**